modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/SortNode.java

```diff
@@ -127,12 +165,32 @@ private void flush() throws Exception {
 
         inLoop = true;
         try {
-            while (requested > 0 && !rows.isEmpty()) {
+            // Prepare final order (reversed).
+            if (limit > 0 && !rows.isEmpty()) {
+                if (reversed == null) {
+                    reversed = new ArrayList<>(rows.size());
+                }
+
+                while (!rows.isEmpty()) {
+                    reversed.add(rows.poll());
```

Comment:
When all data moved from `rows` to `reversed`, then we can free memory and e.g. nullify `rows`.